### PR TITLE
stakepoold: Deprecate wallet RPC ListScripts

### PR DESCRIPTION
**Pull request by JoeGruffins**
Use UserVotingConfig keys and RPC command ValidateAddress instead.

### CODE REVIEW

File: backend/stakepoold/stakepool/stakepool.go [hunk 1]
```diff
@@ -356,15 +356,82 @@ func (spd *Stakepoold) AddMissingTicket(ticketHash []byte) error {
...
+		for _, addr := range addrs {
+			defer wg.Done()
+			validated, err := spd.WalletConnection.RPCClient().ValidateAddress(addr)
+			if err != nil {
+				err = fmt.Errorf("failed to validate address %s", addr.String())
+				log.Error("%s: %v", op, err)
+				results <- &result{err: err}
+				continue
+			}
+			results <- &result{addr: addr.String(), hex: validated.Hex}
```
Comment (jholdstock) [inline backend/stakepoold/stakepool/stakepool.go]: Why is this using `validated.Hex` rather than `validated.Script`?

Comment (JoeGruffins) [reply backend/stakepoold/stakepool/stakepool.go]: iirc the type of script is there, not the script, for example, "multisig"

Comment (JoeGruffins) [reply backend/stakepoold/stakepool/stakepool.go]: https://github.com/decred/dcrwallet/blob/270804fbd664b5cb124400c3ea10dcdb21bb51de/internal/rpc/jsonrpc/methods.go#L3620

**Comment by jrick** [comment]
functionally this looks fine and it works in a pinch when using the most recent wallet but i don't believe this is the best way to to about this.  ideally it would be comparing addresses across wallets, and only looking up the scripts for addresses that some wallet was missing.  getaddressesbyaccount appears broken on the imported account in dcrwallet right now however...

doesn't stakepoold already have a complete list of all addresses it should be watching for ticket purchases? could that be used instead?

**Comment by JoeGruffins** [comment]
@jrick We have to check to make sure that each wallet has all user redeem scripts. Are you saying that there's a better way to check than validateaddress? If so we will use that.

**Comment by jrick** [comment]
there should be, but getaddressesbyaccount seems broken for the imported account currently.  if that worked, you could easily check for all p2sh addresses for each script.

in lieu of that, you can perform validateaddress calls checking to see if each result for the p2sh addresses has `ismine: true`.  if one is not, these can be recorded per-wallet, fetching the from other sources (e.g. other wallets which had them perhaps), and imported.

**Comment by JoeGruffins** [comment]
Created an issue #613 

Also, I just realized that stakepoold works without a connection to the database. I think that needs to be changed.

**Comment by JoeGruffins** [comment]
@jrick 
> in lieu of that, you can perform validateaddress calls checking to see if each result for the p2sh addresses
Are they multisig or p2sh addresses? They are two different things, correct? do multisig need to be followed by a p2sh script to be spent?

**Comment by jrick** [comment]
Currently, the ticket outputs are p2sh. the p2sh redeem script is a 1-of-2 multisig.

**Comment by JoeGruffins** [comment]
@jholdstock @dajohi 
Unless I'm mistaken, master is currently broken without these changes because #609 has been merged?

**Comment by itswisdomagain** [comment]
> there should be, but getaddressesbyaccount seems broken for the imported account currently. if that worked, you could easily check for all p2sh addresses for each script.
> 
> in lieu of that, you can perform validateaddress calls checking to see if each result for the p2sh addresses has `"ismine": true`. if one is not, these can be recorded per-wallet, fetching the script from other sources (e.g. other wallets which had them perhaps), and imported.

dcrstakepool has the list of p2sh addresses it watches, together with the redeem script for each. Calling `validateaddress` for each address per stakepoold instance/wallet would be a lot of rpc calls being made when dcrstakepool starts up, all just to determine which wallet is missing which script, if any.

I don't know if it was discussed before, the possibility of passing multiple addresses to the `validateaddress` rpc method. If that's doable, the number of rpc calls to check for missing scripts would reduce to perhaps 1 per stakepoold instance/wallet.

Might be a better idea to fix `getaddressesbyaccount` for imported accounts and use that instead, to reduce the number of rpc calls to 1 per wallet (excluding calls to import missing scripts).

**Comment by jrick** [comment]
I forgot to update here, but getaddressesbyaccount on the imported account is fixed in dcrwallet master.

**Comment by JoeGruffins** [comment]
will try to solve #613 in another pr, using getaddressesbyaccount.

**Comment by JoeGruffins** [comment]
#619 which uses getaddressesbyaccount should be considered instead of this pr

**Pull request closed without merging**